activemodel/lib/active_model/attribute_methods.rb

```diff
@@ -443,6 +504,11 @@ def method_name(attr_name)
         end
     end
 
+    def initialize(...)
+      self.class.generate_alias_attributes
+      super(...)
+    end
```

Comment:
Shouldn't that be triggered from the same method that defines the normal attribute methods?

I'd rather avoid such method call whenever a model is instantiated, especially since it synchronize a mutex.

Reply:
>especially since it synchronize a mutex.

I forgot to add one more `return` to avoid hitting the `synchronize` if we have already generated alias attributes

Ultimately it mimics what Active Record does to define attributes:
https://github.com/rails/rails/blob/9c94a6bac6122fedb018424c88d677a4c7f35ecf/activerecord/lib/active_record/core.rb#L755

The call chain is `ActiveRecord::Core#initialize` -> `ActiveRecord::Core#init_internals`  -> `self.class.        klass.define_attribute_methods`


>be triggered from the same method that defines the normal attribute methods?

It can be in case of Active Record but it made me realize that Active Model eagerly defines attribute methods right at the `attribute` definition - 
https://github.com/rails/rails/blob/9c94a6bac6122fedb018424c88d677a4c7f35ecf/activemodel/lib/active_model/attributes.rb#L59-L61

I think we should just follow the same pattern, we are just interested in Active Record classes so for pure Active Model classes alias attribute generation can stay synchronous

Reply:
Alright, I modified the code to keep alias attribute methods eagerly defined for Active Model to avoid performance regression and changes to `initialize`. This works fine for our goal since we are mainly interested in Active Record models but it also means that pure Active Model classes will not catch every deprecated scenario. I'll need to think how big of a concern this is